pkg/syncer/obsctlsyncer.go

```diff
@@ -0,0 +1,354 @@
+	"gopkg.in/yaml.v3"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	obsctlContextAPIName = "api"
+)
+
+// ObsctlRulesSyncer implements RulesSyncer interface to sync rules to Observatorium API.
+type ObsctlRulesSyncer struct {
+	ctx             context.Context
+	logger          log.Logger
+	skipClientCheck bool
+	k8s             client.Client
+	namespace       string
+
+	apiURL         string
+	audience       string
+	issuerURL      string
+	managedTenants string
+
+	autoDetectSecretsFn func(ctx context.Context,
+		k8s client.Client,
+		namespace, audience, issuerURL, managedTenants string,
+	) (map[string]*config.OIDCConfig, error)
+
+	c *config.Config
+
+	lokiRulesSetOps      *prometheus.CounterVec
+	promRulesSetOps      *prometheus.CounterVec
+	lokiRulesSetFailures *prometheus.CounterVec
+	promRulesSetFailures *prometheus.CounterVec
+}
+
+func NewObsctlRulesSyncer(
+	ctx context.Context,
+	logger log.Logger,
+	kc client.Client,
+	namespace, apiURL, audience, issuerURL, managedTenants string,
+	reg prometheus.Registerer,
+) *ObsctlRulesSyncer {
+	return &ObsctlRulesSyncer{
+		ctx:            ctx,
+		logger:         logger,
+		k8s:            kc,
+		apiURL:         apiURL,
+		namespace:      namespace,
+		audience:       audience,
+		issuerURL:      issuerURL,
+		managedTenants: managedTenants,
+
+		autoDetectSecretsFn: AutoDetectTenantSecrets,
+
+		lokiRulesSetOps: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_loki_rule_sets_total",
+			Help: "Total number of obsctl set operations for lokiv1/v1beta1 rules.",
+		}, []string{"type", "tenant"}),
+		promRulesSetOps: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_prom_rule_sets_total",
+			Help: "Total number of obsctl set operations for monitoringv1 rules.",
+		}, []string{"tenant"}),
+		lokiRulesSetFailures: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_loki_rule_set_failures_total",
+			Help: "Total number of failed obsctl set operations for lokiv1/v1beta1 rules.",
+		}, []string{"type", "tenant"}),
+		promRulesSetFailures: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "obsctl_reloader_prom_rule_set_failures_total",
+			Help: "Total number of failed obsctl set operations for monitoringv1 rules.",
+		}, []string{"tenant"}),
+	}
+}
+
+func AutoDetectTenantSecrets(
+	ctx context.Context,
+	k8s client.Client,
+	namespace, audience, issuerURL, managedTenants string,
+) (map[string]*config.OIDCConfig, error) {
+	tenantSecret := map[string]*config.OIDCConfig{}
+
+	// List secrets.
+	secret := corev1.SecretList{}
+	if err := k8s.List(ctx, &secret, client.InNamespace(namespace)); err != nil {
+		return nil, err
+	}
+
+	// Filter secrets for configured tenants.
+	configuredTenants := strings.Split(managedTenants, ",")
+	for i := range secret.Items {
+		lbls := secret.Items[i].Labels
+
+		if _, ok := lbls["tenant"]; !ok {
```

Comment:
I wonder should we be more explicit with the label here and make sure it conforms to ? Something like `obsctl.controller.io/tenant` etc

This will need to be documented regardless

Reply:
Yes, I plan to, but with that we'd probably need to migrate a bunch of tenant-secrets. Might be ok to do as a next step?